### PR TITLE
ENH: stats: shared array api support information to generate docs and test skip decorators

### PR DESCRIPTION
**Pull request by mdhaber**
#### Reference issue
Supersedes gh-21933
Toward gh-21255

#### What does this implement/fix?
gh-21255 proposes that there should be a central source of information about the Array API capabilities of a given function, and this information should be used to inject documentation. I also thought it would be nice to use this information to control test skips, and it looks like that was also suggested in https://github.com/scipy/scipy/issues/21255#issuecomment-2328382749. This PR implements these ideas.

#### Additional information
Didn't think too carefully about function names, so feel free to suggest specific new ones.

### CODE REVIEW

**Comment by lucascolley** [comment]
> I ran out of steam for using the information to skip tests, so that part is sloppy. We need to augment the way the information is entered and stored (e.g. so we can store skip reasons), and we'll probably want to use the information to produce a single decorator that itself applies a sequence of `pytest.mark.skip_xp_backend` decorators to a function/class - then this decorator gets applied to the test class/function.

any thoughts here @crusaderky ?

File: scipy/_lib/_array_api.py [hunk 1]
```diff
@@ -652,3 +657,91 @@ def xp_default_dtype(xp):
...
+        for attr in ['cpu', 'gpu', 'mask']:
+            val = "✓" if getattr(backend, attr) else " "
+            val += " " * (len('{numpy.}') + len(attr) - 1)
+            setattr(backend, attr, val)
+
+    table = f"""                
+    +---------+-------------+-------------+--------------+
+    | Library | CPU         | GPU         | ``marray``   |
+    +=========+=============+=============+==============+
+    | NumPy   | {numpy.cpu} | {numpy.gpu} | {numpy.mask} |
```
Comment (mdhaber) [inline scipy/_lib/_array_api.py:709]: Looks like Sphinx wants borders between lines to render this as intended

File: scipy/stats/_stats_py.py [hunk 2]
```diff
@@ -4363,6 +4366,10 @@ def confidence_interval(self, confidence_level=0.95, method=None):
         return ci
 
 
+xp_capabilities_table['pearsonr'] = dict(cpu_only=True)
+
+
+@xp_capabilities(**xp_capabilities_table['pearsonr'])
```
Comment (mdhaber) [inline scipy/stats/_stats_py.py:4372]: Assuming we want to leave `skip_xp_backends` right now (I don't want to touch it, anyway), I'm thinking this can accept a sequence of dictionaries, each of which defines valid keargs input to `skip_xp_backends`. We process these to produce the capabilities table here, and it would then be very easy to chain together a test decorator from them since each element is valid input to `skip_xp_backends`.

**Comment by ev-br** [comment]
Tangentially related: I've a prototype of a small hack which records, for a test run, for each scipy function all array types (== backends) it was called with. So that after `python dev.py test -b all` we'll have a list of dicts scipy_function: list_of_backends.
It's not PR ready yet, will clean it up soon (tm).

File: scipy/stats/_stats_py.py [hunk 3]
```diff
@@ -4363,6 +4366,10 @@ def confidence_interval(self, confidence_level=0.95, method=None):
         return ci
 
 
+xp_capabilities_table['pearsonr'] = dict(cpu_only=True)
```
Comment (mdhaber) [inline scipy/stats/_stats_py.py:4369]: Currently, `xp_capabilities_table` is a dictionary defined in the module per Ralf's suggestion, but it can be defined anywhere.

The keys are function names, and the values are dictionaries that accept keyword input similar to the `skip_xp_backend` decorator, like `cpu_only`, `np_only`, and `exceptions`. The one difference is that all the backends to skip and corresponding reasons would be provided as list of tuples, e.g.
```python3
[('cupy', 'I said so'), 
 ('jax.numpy', 'Just because']
```
Or I suppose these tuple could be provided as `*args` (each tuple is an element), but currently it's a single argument, `skip_backends`.

I've only implemented `skip_backends` and `cpu_only` for now.

File: scipy/stats/_stats_py.py [hunk 4]
```diff
@@ -4363,6 +4366,10 @@ def confidence_interval(self, confidence_level=0.95, method=None):
...
+
+
+@xp_capabilities(xp_capabilities_table)
```
Comment (mdhaber) [inline scipy/stats/_stats_py.py:4372]: The job of the decorator that gets applied to the function is to add the table to the documentation. The decorator factory `xp_capabilities` accepts the module's `xp_capabilities_table` as an argument. The decorator returned by the factory gets passed the function as an argument `f`, so inside that we can look up `xp_capabilities_table[f.__name__]`.

File: scipy/stats/tests/test_stats.py [hunk 5]
```diff
@@ -379,7 +381,7 @@ def test_pROUNDROUND(self):
         assert_approx_equal(r,1.0)
 
 
-@skip_xp_backends(cpu_only=True)
+@make_skip_xp_backends('pearsonr', xp_capabilities_table)
```
Comment (mdhaber) [inline scipy/stats/tests/test_stats.py:384]: Now we only have to apply one decorator to test classes. The `make_skip_xp_backends` decorator factory looks up the information in `xp_capabilities_table['pearsonr']`, creates all the appropriate decorators with that information, then returns a single decorator that chains them all together.

If we were to name test classes in a standard way (which is usually the case right now, especially as we go back through and do arrray API translations), we wouldn't have to provide the function name as an argument.

**Review by mdhaber — COMMENTED** [review]
OK, I went ahead and implemented what I had in mind for skipping tests based on the shared info.

File: scipy/_lib/_array_api.py [hunk 6]
```diff
@@ -652,3 +657,115 @@ def xp_default_dtype(xp):
...
+def _make_capabilities_note(fun_name, capabilities):
+    table = _make_capabilities_table(capabilities)
+    note = f"""
+    `{fun_name}` has experimental support for Python Array API Standard compatible
+    backends in addition to NumPy. Please consider testing these features
+    by setting an environment variable ``SCIPY_ARRAY_API=1`` and providing
+    CuPy, PyTorch, JAX, or Dask arrays as array arguments. The following
+    combinations of backend and device (or other capability) are supported.
+    {table}
+     See :ref:`dev-arrayapi` for more information.
```
Comment (mdhaber) [inline scipy/_lib/_array_api.py:732]:
```suggestion
    See :ref:`dev-arrayapi` for more information.
```
Other than this, looks pretty good to me:
image

File: scipy/_lib/_array_api.py [hunk 7]
```diff
@@ -652,3 +657,115 @@ def xp_default_dtype(xp):
...
+        capabilities = _make_capabilities(skip_backends=skip_backends,
+                                          cpu_only=cpu_only)
+
+        @functools.wraps(f)
+        def wrapper(*args, **kwargs):
+            return f(*args, **kwargs)
+
+        note = _make_capabilities_note(fun_name, capabilities)
+        doc = FunctionDoc(wrapper)
+        doc['Notes'].append(note)
```
Comment (j-bowhay) [inline scipy/_lib/_array_api.py:751]: Does `numpydoc` allow us define a new section like `Array API Compatibility` for better visibility rather than being at the end of the notes?

File: scipy/_lib/_array_api.py [hunk 8]
```diff
@@ -652,3 +657,115 @@ def xp_default_dtype(xp):
...
+
+    for backend in [numpy, cupy, torch, jax, dask]:
+        for attr in ['cpu', 'gpu', 'mask']:
+            val = "✓" if getattr(backend, attr) else " "
+            val += " " * (len('{numpy.}') + len(attr) - 1)
+            setattr(backend, attr, val)
+
+    table = f"""                
+    +---------+-------------+-------------+--------------+
+    | Library | CPU         | GPU         | ``marray``   |
```
Comment (j-bowhay) [inline scipy/_lib/_array_api.py:707]: Certainly not required for this pr but it could be nice in the future if the `marray` text and other libraries linked to their docs

File: scipy/stats/_stats_py.py (hunk 4 above)
Comment (j-bowhay) [reply scipy/stats/_stats_py.py:4372]: There was some concern when I originally suggested the idea that using decorators to inject the docs might cause too much of a slowdown at import time if there are lots of functions. However it seems reasonable to me to start with something like this and then, if this becomes an issue, move to some sphinx magic

File: scipy/stats/tests/test_stats.py (hunk 5 above)
Comment (j-bowhay) [reply scipy/stats/tests/test_stats.py:384]: If we defined `xp_capabilities_table` centrally in `_array_api.py` then the decorators could just use it directly?

File: scipy/stats/_stats_py.py (hunk 3 above)
Comment (j-bowhay) [reply scipy/stats/_stats_py.py:4369]: Looks like we'd need mark `jax.numpy` support as false since all tests are skipped due to item assignment so a good opportunity to test this out

**Review by j-bowhay — COMMENTED** [review]
Thanks for getting the ball rolling on this, I think it's an important step in making the array API support more useable!

File: scipy/_lib/_array_api.py (hunk 7 above)
Comment (mdhaber) [reply scipy/_lib/_array_api.py:751]: I'm not sure, but if not, we could ask.

File: scipy/_lib/_array_api.py (hunk 8 above)
Comment (mdhaber) [reply scipy/_lib/_array_api.py:707]: Yeah I thought about that, but I'm not sure reST tables support links. I tried citations before, and that didn't work.

File: scipy/_lib/_array_api.py (hunk 7 above)
Comment (j-bowhay) [reply scipy/_lib/_array_api.py:751]: Not required for this pr though but it would be nice for the future

File: scipy/stats/_stats_py.py (hunk 4 above)
Comment (mdhaber) [reply scipy/stats/_stats_py.py:4372]: Yeah. But if that comes up again, then it's basically saying that only rendered docs matter. If that's the case, then we should be "allowed" to use :math: liberally.

File: scipy/stats/tests/test_stats.py (hunk 5 above)
Comment (mdhaber) [reply scipy/stats/tests/test_stats.py:384]: They could, but didn't we want to keep it close to the functions?

Comment (mdhaber) [reply scipy/stats/tests/test_stats.py:384]: > I mean just defining the empty dictionary centrally

Ah, yes, thought about that, but shied away given the new multithreading work. I would definitely like that better than separate dictionaries in each module, though. It would mostly eliminate the advantages of attaching the information to the function. Is it ok to have a global dictionary that's mutated all over SciPy (but only once, on import)?

File: doc/source/dev/api-dev/array_api.rst [hunk 9]
```diff
@@ -121,6 +121,32 @@ Support is provided in `scipy.stats` for the following functions:
...
+| NumPy   | ✓           | ✗           | ✓            |
++---------+-------------+-------------+--------------+
+| CuPy    | ✗           | ✓           |              |
++---------+-------------+-------------+--------------+
+| PyTorch | ✓           |             |              |
++---------+-------------+-------------+--------------+
+| JAX     | ✓           | ✓           |              |
++---------+-------------+-------------+--------------+
+| Dask    |             |             |              |
++---------+-------------+-------------+--------------+
```
Comment (mdhaber) [inline doc/source/dev/api-dev/array_api.rst:138]: This looks like it's aligned locally.

<details>

image

</details>

**Review by mdhaber — COMMENTED** [review]
I defined the dictionary in `_array_api.py` as suggested. If that's going to be a problem, we can change it to having one dictionary per module.

I added a bit of an explanation of the table to `array_api.rst`; that links to `marray`. We could add links to the other backend websites there, too.

I think the machinery now supports all the `skip_xp_backends` features that we use in the tests - besides skipping individual backends, that's `cpu_only`, possibly with `exceptions`, and `np_only`.  If `exceptions` is supposed to work with `np_only`, I'd suggest we trim things down a bit so that there is not so much variety in how that can be accomplished. 

I also applied this work to `zmap`/`zscore`. If this is looking good, I can apply it to the rest of `_stats_py.py` and tests; I leave the rest of SciPy to follow-up work.

File: doc/source/dev/api-dev/array_api.rst [hunk 10]
```diff
@@ -121,6 +121,32 @@ Support is provided in `scipy.stats` for the following functions:
 `scipy.stats.jarque_bera`, `scipy.stats.bartlett`, `scipy.stats.power_divergence`,
 and `scipy.stats.monte_carlo_test`.
 
+Some features provide a capability table in the documentation like this:
+
++---------+-------------+-------------+--------------+
+| Library | CPU         | GPU         | ``marray``   |
++=========+=============+=============+==============+
+| NumPy   | ✓           | ✗           | ✓            |
```
Comment (rgommers) [inline doc/source/dev/api-dev/array_api.rst:129]: minor: probably better to use `x` to indicate support is missing when that's possible, and "n/a" when the cell can never be checked - NumPy on GPU or CuPy on CPU will never be a thing.

**Comment by rgommers** [comment]
I had a quick look (in transit, only a bit of time now) and it looks promising to me. Thanks for working on this @mdhaber!

The `@make_skip_...` approach to reuse the compatibility info LGTM. There will still be cases where `@skip_...`will be needed, when the incompatibility is in the test itself instead of in the function under test, but that should be a minority of cases. Overall this should make things more consistent and a bit more concise inside the test files.

File: scipy/stats/_stats_py.py [hunk 11]
```diff
@@ -4363,6 +4371,10 @@ def confidence_interval(self, confidence_level=0.95, method=None):
         return ci
 
 
+xp_capabilities_table['pearsonr'] = dict(cpu_only=True, exceptions=['cupy'])
```
Comment (j-bowhay) [inline scipy/stats/_stats_py.py:4374]: I'm not sure I have a strong opinion but I wonder if we should should use the function handle as the key instead as, for example, `linalg` and `sparse.linalg` have functions with the same name

Comment (mdhaber) [reply scipy/stats/_stats_py.py:4374]: The other example shows why we don't necessarily just want one key per function, and in that case strings are appropriate. Collisions will be caught during testing, and the user is welcome to fix them by providing a fully qualifiied name as a string or function handle, if they prefer.

Comment (j-bowhay) [reply scipy/stats/_stats_py.py:4374]: Fair point. Another advantage of using the function handle is that it would allow for a module level comparability table in the docs but I guess we can always change things down the line if that's something we want to add

File: scipy/_lib/_array_api.py [hunk 12]
```diff
@@ -652,3 +657,139 @@ def xp_default_dtype(xp):
...
+    +=========+=============+=============+==============+
+    | NumPy   | {numpy.cpu} | ✗           | {numpy.mask} |
+    +---------+-------------+-------------+--------------+
+    | CuPy    | ✗           | {cupy.gpu } | {cupy.mask } |
+    +---------+-------------+-------------+--------------+
+    | PyTorch | {torch.cpu} | {torch.gpu} | {torch.mask} |
+    +---------+-------------+-------------+--------------+
+    | JAX     | {jax.cpu  } | {jax.gpu  } | {jax.mask  } |
```
Comment (crusaderky) [inline scipy/_lib/_array_api.py:724]: I think it's important to distinguish between JAX in eager mode vs. jax.jit

File: scipy/_lib/_array_api.py [hunk 13]
```diff
@@ -652,3 +657,139 @@ def xp_default_dtype(xp):
...
+    +---------+-------------+-------------+--------------+
+    | Dask    | {dask.cpu } | {dask.gpu } | {dask.mask } |
```
Comment (crusaderky) [inline scipy/_lib/_array_api.py:726]: Likewise, saying "it functionally reaches the end with dask" is a very different thing from "it is usable in production with Dask and you should expect OK performance at least in some cases" due to materializations.

This and the above information about JAX is currently stored in the `lazy_xp_function` helpers in the test modules.

File: scipy/stats/_stats_py.py [hunk 14]
```diff
@@ -4363,6 +4371,10 @@ def confidence_interval(self, confidence_level=0.95, method=None):
...
+
+
+@xp_capabilities()
```
Comment (crusaderky) [inline scipy/stats/_stats_py.py:4377]: UX nit: the `@xp_capabilities` decorator could inject into `xp_capabilities_table` itself
```suggestion
@xp_capabilities(cpu_only=True, exceptions=['cupy'])
```

File: scipy/stats/_stats_py.py [hunk 15]
```diff
@@ -2660,6 +2660,13 @@ def _isconst(x):
         return (y[0] == y).all(keepdims=True)
 
 
+xp_capabilities_table['zmap_zscore'] = dict(skip_backends=[
```
[the remark this answers is not in the document]
Comment (ev-br) [reply scipy/stats/_stats_py.py:2663]: +1. `stats.zmap_score` would be better, really.

File: scipy/_lib/_array_api.py (hunk 12 above)
Comment (mdhaber) [reply scipy/_lib/_array_api.py:724]: Would there be a column or row for that? I was aware of the distinction, but I don't know much more.

File: scipy/stats/_stats_py.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (mdhaber) [reply scipy/stats/_stats_py.py:2663]: No. There is not a function with that name.

File: scipy/_lib/_array_api.py (hunk 12 above)
Comment (crusaderky) [reply scipy/_lib/_array_api.py:724]: jax.jit functionality is always a subset of eager jax. So maybe the layout could look something like this?
```
JAX | {green check mark} |
JAX | {yellow warning symbol} Eager only |
JAX | {red x} |
```

File: scipy/stats/_stats_py.py (hunk 14 above)
Comment (mdhaber) [reply scipy/stats/_stats_py.py:4377]: I actually described that it would work that way before writing it. Not sure why I didn't do that, so I'll change it.

File: scipy/_lib/_array_api.py (hunk 12 above)
Comment (mdhaber) [reply scipy/_lib/_array_api.py:724]: Sure. I was hoping to start simple, though. Would this require a change to the information stored in the `xp_capabilities_table ` dictionary, or do you mean below that it be looked up for a given function with `lazy_xp_function`?

**Review by j-bowhay — APPROVED** [review]
A big improvement IMO for visibility and for figuring out backend support. I'm sure we will end up tweaking things and finalizing the design (eg. https://github.com/scipy/scipy/pull/22609#discussion_r1980342136) but I don't think we need to fully converge here so I would propose merging soon.

Looks like we have picked up a merge conflict in the meantime.

File: scipy/stats/tests/test_stats.py [hunk 16]
```diff
@@ -6313,8 +6312,7 @@ def test_ttest_1samp_new_omit(xp):
     xp_assert_close(t, tr)
 
 
-@make_skip_xp_backends('ttests')
-@pytest.mark.skip_xp_backends('jax.numpy', reason='Generic impl mutates array.')
+@pytest.mark.skip_xp_backends('jax.numpy', reason='Generic stdtrit mutates array.')
```
Comment (mdhaber) [inline scipy/stats/tests/test_stats.py:6315]:
```suggestion
@make_skip_xp_backends('ttests')
@pytest.mark.skip_xp_backends('jax.numpy', reason='Generic stdtrit mutates array.')
```

File: scipy/stats/tests/test_stats.py [hunk 17]
```diff
@@ -75,7 +75,6 @@
 ROUND = array([0.5,1.5,2.5,3.5,4.5,5.5,6.5,7.5,8.5], float)
 
 
-@make_skip_xp_backends("trimmed statistics")
 class TestTrimmedStats:
```
Comment (mdhaber) [inline scipy/stats/tests/test_stats.py:78]:
```suggestion
@make_skip_xp_backends("trimmed statistics")
class TestTrimmedStats:
````

**Comment by mdhaber** [comment]
Merge conflicts resolved @j-bowhay. Thanks!
